src/libopensc/card-starcos.c

```diff
@@ -215,6 +251,182 @@ static int process_fci(sc_context_t *ctx, sc_file_t *file,
 	return SC_SUCCESS;
 }
 
+static int process_fci_v3_4(sc_context_t *ctx, sc_file_t *file,
+		       const u8 *buf, size_t buflen)
+{
+	size_t taglen, len = buflen;
+	const u8 *tag = NULL, *p;
+
+	sc_debug(ctx, SC_LOG_DEBUG_NORMAL, "processing %d FCI bytes\n", buflen);
+
+	if (buflen < 2)
+		return SC_ERROR_INTERNAL;
+	if (buf[0] != 0x6f)
+		return SC_ERROR_INVALID_DATA;
+	len = (size_t)buf[1];
+	if (buflen - 2 < len)
+		return SC_ERROR_INVALID_DATA;
+
+	/* defaults */
+	file->type = SC_FILE_TYPE_WORKING_EF;
+	if (len == 0) {
+		SC_FUNC_RETURN(ctx, 2, SC_SUCCESS);
+	}
+
+	p = buf + 2;
+	file->ef_structure = SC_FILE_TYPE_DF;
+	file->shareable = 1;
+	tag = sc_asn1_find_tag(ctx, p, len, 0x84, &taglen);
+	if (tag != NULL && taglen > 0 && taglen <= 16) {
+		memcpy(file->name, tag, taglen);
+		file->namelen = taglen;
+		sc_debug(ctx, SC_LOG_DEBUG_NORMAL, "filename %s",
+			sc_dump_hex(file->name, file->namelen));
+	}
+	return SC_SUCCESS;
+}
+
+static int process_fcp_v3_4(sc_context_t *ctx, sc_file_t *file,
+		       const u8 *buf, size_t buflen)
+{
+	size_t taglen, len = buflen;
+	const u8 *tag = NULL, *p;
+
+	sc_debug(ctx, SC_LOG_DEBUG_NORMAL, "processing %d FCP bytes\n", buflen);
+
+	if (buflen < 2)
+		return SC_ERROR_INTERNAL;
+	if (buf[0] != 0x62)
+		return SC_ERROR_INVALID_DATA;
+	len = (size_t)buf[1];
+	if (buflen - 2 < len)
+		return SC_ERROR_INVALID_DATA;
+	p = buf + 2;
+
+	tag = sc_asn1_find_tag(ctx, p, len, 0x80, &taglen);
```

Comment:
I would love to give these numbers some meaningful defines, but that's difficult without documentation of the APDUs. The code is based on the STARCOS 3 patch from Martin Vogt and some trial-and-error to determine the correct numbers for STARCOS 3.4.

Besides, the original STARCOS 2.3/2.4 code also has lots of magic numbers everywhere :-)